[IOTDB-220]Add hot-load configuration function

fanhualta opened the pull request:
This is a pr to add hot-load configuration function which  can modify configurations without restarting IoTDB system. 

The jira link: https://issues.apache.org/jira/projects/IOTDB/issues/IOTDB-220

liutaohua commented on server/src/main/java/org/apache/iotdb/db/qp/executor/QueryProcessExecutor.java:
```diff
@@ -130,13 +130,16 @@ public boolean processNonQuery(PhysicalPlan plan) throws ProcessorException {
         operateTTL((SetTTLPlan) plan);
         return true;
+      case LOAD_CONFIGURATION:
+        IoTDBDescriptor.getInstance().loadHotModifiedProps();
```
if some param is error ,some param is true , the load is dirty.
for example:
default pararm is :
`
tsfile_size_threshold = 1000
group_size_in_byte = 300
`
i will change it to 
`
tsfile_size_threshold = 100
group_size_in_byte = 30**f**
`
it look's like tsfile_size_threshold will be start work

fanhualta replied:
These parameters need to be updated by the user, so I will update all parameters until I encounter a wrong parameter. In the scenario you described, is there any problem? Equivalent to that the user only updated `tsfile_size_threshold = 100`.

liutaohua replied:
write  to disk size will all allways max than `tsfile_size_threshold`

fanhualta replied:
I don't think so, `tsfile_size_threshold` is just a lower bound to close the tsfile. So if `tsfile_size_threshold` is smaller than `group_size_in_byte`, it just means you can close the tsfile after you flush one memtable, which is acceptable in IoTDB and will not crash the system.

liutaohua replied:
yes, i haven't thought of a scene that could cause a crash, but if there are multiple interrelated parameters， may be have some hidden bugs.

mdf369 commented on server/src/main/java/org/apache/iotdb/db/conf/IoTDBDescriptor.java:
```diff
@@ -160,14 +151,6 @@ private void loadProps() {
-      conf.setForceWalPeriodInMs(Long
-          .parseLong(properties.getProperty("force_wal_period_in_ms",
-              Long.toString(conf.getForceWalPeriodInMs()))));
-
```
Why no longer need to set these configs?

fanhualta replied:
I extract them into a separate method called `loadWALProps();` which will reduce duplicate codes because these parameters could be hot-loaded.

mdf369 commented on server/src/main/java/org/apache/iotdb/db/conf/IoTDBDescriptor.java:
```diff
@@ -353,58 +313,159 @@ private void loadProps() {
-      try {
-        inputStream.close();
-      } catch (IOException e) {
-        logger.error("Fail to close config file input stream because ", e);
```
By putting the initialization of InputStream into try statement, will it automatically close the stream and catch the IOException during closing?

fanhualta replied:
Yes

mdf369 commented on server/src/main/java/org/apache/iotdb/db/conf/directories/DirectoryManager.java:
```diff
@@ -64,11 +65,53 @@ private DirectoryManager() {
+  public void updateFileFolders() throws LoadConfigurationException {
+    try {
+      List<String> sequenceFileFolders =
+          new ArrayList<>(Arrays.asList(IoTDBDescriptor.getInstance().getConfig().getDataDirs()));
```
Why not `List<String> sequenceFileFolders =
          Arrays.asList(IoTDBDescriptor.getInstance().getConfig().getDataDirs());`?

fanhualta replied:
I just imitate line 47 in this class. I have fixed them all.

mdf369 commented on server/src/main/java/org/apache/iotdb/db/conf/directories/strategy/DirectoryStrategy.java:
```diff
@@ -33,19 +35,17 @@
 
   /**
-   * All the folders of data files, should be init once the subclass is created.
+   * All the folders of data files, should be setFolders once the subclass is created.
```
"setFolders() should be called"

fanhualta replied:
fixed

mdf369 commented on server/src/main/java/org/apache/iotdb/db/conf/directories/strategy/DirectoryStrategy.java:
```diff
@@ -33,19 +35,17 @@
 
   /**
-   * To init folders. Do not recommend to overwrite.
+   * To setFolders folders. Do not recommend to overwrite.
```
Same as above

fanhualta replied:
fixed.

mdf369 commented on server/src/main/java/org/apache/iotdb/db/conf/directories/strategy/DirectoryStrategy.java:
```diff
@@ -33,19 +35,17 @@
+   * All the folders of data files, should be setFolders once the subclass is created.
    */
-  List<String> folders;
+  List<String> folders = new ArrayList<>();
```
In normal situations, this new ArrayList is useless, is this initialization necessary?

fanhualta replied:
Just in case of `NullPointer`. which may be unnecessary. I have removed it.

mdf369 commented on server/src/main/java/org/apache/iotdb/db/engine/querycontext/ReadOnlyMemChunk.java:
```diff
@@ -46,7 +46,7 @@
 
   /**
-   * init by TSDataType and TimeValuePairSorter.
+   * setFolders by TSDataType and TimeValuePairSorter.
```
grammar

fanhualta replied:
I have fixed them all.

mdf369 commented on server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java:
```diff
@@ -358,7 +358,7 @@ public void insert(InsertPlan insertPlan) throws QueryProcessException {
     writeLock();
     try {
-      // init map
+      // setFolders map
```
grammar

mdf369 commented on server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java:
```diff
@@ -373,7 +373,7 @@ public void insert(InsertPlan insertPlan) throws QueryProcessException {
     writeLock();
     try {
-      // init map
+      // setFolders map
```
grammar

mdf369 commented on server/src/main/java/org/apache/iotdb/db/qp/logical/crud/SFWOperator.java:
```diff
@@ -51,7 +51,7 @@ public SelectOperator getSelectOperator() {
 
   /**
-   * set selectOperator, then init hasAggregation according to selectOperator.
+   * set selectOperator, then setFolders hasAggregation according to selectOperator.
```
grammar

mdf369 commented on server/src/main/java/org/apache/iotdb/db/qp/logical/crud/SelectOperator.java:
```diff
@@ -32,7 +32,7 @@
 
   /**
-   * init with tokenIntType, default operatorType is <code>OperatorType.SELECT</code>.
+   * setFolders with tokenIntType, default operatorType is <code>OperatorType.SELECT</code>.
```
grammar

mdf369 commented on server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByEngineDataSet.java:
```diff
@@ -60,7 +60,7 @@ public GroupByEngineDataSet(long jobId, List<Path> paths, long unit, long origin
     this.functions = new ArrayList<>();
 
-    // init group by time partition
+    // setFolders group by time partition
```
grammar

mdf369 commented on server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithValueFilterDataSet.java:
```diff
@@ -65,7 +65,7 @@ public GroupByWithValueFilterDataSet(long jobId, List<Path> paths, long unit, lo
 
   /**
-   * init reader and aggregate function.
+   * setFolders reader and aggregate function.
```
grammar

mdf369 commented on server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java:
```diff
@@ -70,12 +70,12 @@ public GroupByWithoutValueFilterDataSet(long jobId, List<Path> paths, long unit,
 
   /**
-   * init reader and aggregate function.
+   * setFolders reader and aggregate function.
```
grammar

mdf369 commented on server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java:
```diff
@@ -70,12 +70,12 @@ public GroupByWithoutValueFilterDataSet(long jobId, List<Path> paths, long unit,
       throws StorageEngineException, PathException, IOException {
     initAggreFuction(aggres);
-    // init reader
+    // setFolders reader
```
grammar

mdf369 commented on server/src/main/java/org/apache/iotdb/db/query/externalsort/SimpleExternalSortEngine.java:
```diff
@@ -132,7 +132,7 @@ public void setMinExternalSortSourceCount(int minExternalSortSourceCount) {
 
    /**
-    * init IPointReader with ChunkReaderWrap.
+    * setFolders IPointReader with ChunkReaderWrap.
```
grammar

mdf369 commented on server/src/main/java/org/apache/iotdb/db/query/externalsort/SimpleExternalSortEngine.java:
```diff
@@ -144,7 +144,7 @@ public void setMinExternalSortSourceCount(int minExternalSortSourceCount) {
 
    /**
-    * init IReaderByTimestamp with ChunkReaderWrap.
+    * setFolders IReaderByTimestamp with ChunkReaderWrap.
```
grammar

mdf369 commented on server/src/main/java/org/apache/iotdb/db/query/reader/fileRelated/UnSealedTsFileIterateReader.java:
```diff
@@ -109,7 +109,7 @@ private IAggregateReader initUnSealedTsFileDiskReader(TsFileResource unSealedTsF
 
     if (metaDataList == null || metaDataList.isEmpty()) {
-      // init fileSeriesReader
+      // setFolders fileSeriesReader
```
grammar

mdf369 commented on server/src/main/java/org/apache/iotdb/db/query/reader/fileRelated/UnSealedTsFileIterateReader.java:
```diff
@@ -126,7 +126,7 @@ private IAggregateReader initUnSealedTsFileDiskReader(TsFileResource unSealedTsF
           .get(unSealedTsFile, false);
       IChunkLoader chunkLoader = new ChunkLoaderImpl(unClosedTsFileReader);
-      // init fileSeriesReader
+      // setFolders fileSeriesReader
```
grammar

mdf369 commented on server/src/main/java/org/apache/iotdb/db/query/reader/resourceRelated/SeqResourceIterateReader.java:
```diff
@@ -174,7 +174,7 @@ private IAggregateReader initSealedTsFileReader(TsFileResource sealedTsFile, Fil
     IChunkLoader chunkLoader = new ChunkLoaderImpl(tsFileReader);
 
-    // init fileSeriesReader
+    // setFolders fileSeriesReader
```
grammar

mdf369 commented on server/src/main/java/org/apache/iotdb/db/sync/receiver/transfer/SyncServiceImpl.java:
```diff
@@ -143,7 +143,7 @@ public SyncStatus init(String storageGroup) {
       syncLog.get().startSyncDeletedFilesName();
     } catch (IOException e) {
-      logger.error("Can not init sync process", e);
+      logger.error("Can not setFolders sync process", e);
```
grammar

mdf369 commented on server/src/main/java/org/apache/iotdb/db/sync/receiver/transfer/SyncServiceImpl.java:
```diff
@@ -185,9 +185,9 @@ public SyncStatus initSyncData(String filename) throws TException {
+      logger.error("Can not setFolders sync resource for file {}", filename, e);
       return getErrorResult(
-          String.format("Can not init sync resource for file %s because %s", filename,
+          String.format("Can not setFolders sync resource for file %s because %s", filename,
```
grammar

mdf369 commented on server/src/test/java/org/apache/iotdb/db/query/reader/universal/FakedSeriesReader.java:
```diff
@@ -31,12 +31,12 @@
   private final static TSDataType DATA_TYPE = TSDataType.INT64;
 
-  // init with time list and value
+  // setFolders with time list and value
```
grammar

mdf369 commented on server/src/test/java/org/apache/iotdb/db/query/reader/universal/FakedSeriesReader.java:
```diff
@@ -31,12 +31,12 @@
 
 
-  // init with startTime, size, interval and modValue
+  // setFolders with startTime, size, interval and modValue
```
grammar

LeiRui commented on docs/Documentation-CHN/UserGuide/3-Server/4-Config Manual.md:
````diff
@@ -25,15 +25,20 @@
+
+触发方式：客户端发送```load configuration```命令至IoTDB Server，客户端的使用方式详见第4章
 
 ### 环境配置项
````
The corresponding content in English is as follows:
![image](https://user-images.githubusercontent.com/33376433/68764837-6afad900-0656-11ea-8225-c305e3214dd8.png)
So, I think the `Config Manual` is the highest level and should better be interpreted as `配置手册`. And the `Hot Modification Configuration` may appear at the wrong level.

fanhualta replied:
Good comment. I have corrected them.

LeiRui commented:
Others
- `MAX_HEAP_SIZE` or `HEAP_NEWSIZE` are described in the `环境配置项` of `4-Config Manual.md`, but I can't find them in `iotdb-env.bat`.
- In the "系统配置项" of `4-Config Manual.md`, `tsfile-format.properties` is still referenced, but isn't it already merged into `iotdb-engine.properties`?
- `back_loop_period` in the `Documentation-CHN/UserGuide/3-Server/4-Config Manual.md` should be `back_loop_period_in_second`.

fanhualta commented on docs/Documentation/UserGuide/3-Server/4-Config Manual.md:
```diff
@@ -29,13 +29,22 @@ Before starting to use IoTDB, you need to config the configuration files first.
+In the parameters described below, these parameters whose way of `Effective` is `trigger` support hot modification.
+
+Trigger way: The client sends the command `load configuration` to the IoTDB server. See Chapter 4 for the usage of the client.
+为方便用户使用，IoTDB Server为用户提供了热修改功能，即在系统运行过程中修改`iotdb-engine.properties`中部分配置参数并即时应用到系统中。下面介绍的参数中，改后
```
It's my carelessness. I have removed them.

fanhualta commented on docs/Documentation-CHN/UserGuide/3-Server/4-Config Manual.md:
```diff
@@ -25,15 +25,20 @@
+## 热修改配置项
+
+为方便用户使用，IoTDB Server为用户提供了热修改功能，即在系统运行过程中修改`iotdb-engine.properties`中部分配置参数并即时应用到系统中。下面介绍的参数中，改后
+生效方式为`触发修改`的均为支持热修改的配置参数。
```
I have fixed it.

fanhualta commented:
> Others
> 
> * `MAX_HEAP_SIZE` or `HEAP_NEWSIZE` are described in the `环境配置项` of `4-Config Manual.md`, but I can't find them in `iotdb-env.bat`.
> * In the "系统配置项" of `4-Config Manual.md`, `tsfile-format.properties` is still referenced, but isn't it already merged into `iotdb-engine.properties`?
> * `back_loop_period` in the `Documentation-CHN/UserGuide/3-Server/4-Config Manual.md` should be `back_loop_period_in_second`.

1. I have added these two parameters in `iotdb-env.bat` and test in the windows environment.
2. `tsfile-format.properties` is already merged into `iotdb-engine.properties` and I have removed its reference.
3. I have corrected its name.
